fv3core/stencils/xtp_u.py

```diff
@@ -16,82 +16,126 @@
     else:
-        tmp = 1.0
-    return xppm.final_flux(courant, u, fx0, tmp)
+        advection_mask = 1.0
+    return xppm.apply_flux(ub_contra, u, fx0, advection_mask)
 
 
-def _xtp_u(
+def xtp_u_stencil_defn(
     courant: FloatField,
     u: FloatField,
     flux: FloatField,
     dx: FloatFieldIJ,
     dxa: FloatFieldIJ,
     rdx: FloatFieldIJ,
 ):
+    with computation(PARALLEL), interval(...):
+        flux = advect_u_along_x(courant, u, dx, dxa, rdx, 1.0)
+
+
+@gtscript.function
+def get_bl_br(u, dx, dxa):
+    """
+    Args:
+        u: covariant x-wind on D-grid
+        dx: gridcell spacing in x-direction
+        dxa: gridcell spacing in x-direction on A-grid
+
+    Returns:
+        bl: ???
+        br: ???
+    """
     from __externals__ import i_end, i_start, iord, j_end, j_start
 
-    with computation(PARALLEL), interval(...):
+    if __INLINED(iord < 8):
+        u_on_cell_corners = xppm.compute_al(u, dx)
 
-        if __INLINED(iord < 8):
-            al = xppm.compute_al(u, dx)
+        bl = u_on_cell_corners[0, 0, 0] - u[0, 0, 0]
+        br = u_on_cell_corners[1, 0, 0] - u[0, 0, 0]
 
-            bl = al[0, 0, 0] - u[0, 0, 0]
-            br = al[1, 0, 0] - u[0, 0, 0]
+    else:
+        dm = xppm.dm_iord8plus(u)
+        u_on_cell_corners = xppm.al_iord8plus(u, dm)
+
+        compile_assert(iord == 8)
+
+        bl, br = xppm.blbr_iord8(u, u_on_cell_corners, dm)
+        bl, br = xppm.bl_br_edges(bl, br, u, dxa, u_on_cell_corners, dm)
 
-        else:
-            dm = xppm.dm_iord8plus(u)
-            al = xppm.al_iord8plus(u, dm)
+        with horizontal(region[i_start + 1, :], region[i_end - 1, :]):
+            bl, br = yppm.pert_ppm_standard_constraint_fcn(u, bl, br)
 
-            compile_assert(iord == 8)
+    # Zero corners
+    with horizontal(
+        region[i_start - 1 : i_start + 1, j_start],
+        region[i_start - 1 : i_start + 1, j_end + 1],
+        region[i_end : i_end + 2, j_start],
+        region[i_end : i_end + 2, j_end + 1],
+    ):
+        bl = 0.0
+        br = 0.0
+    return bl, br
 
-            bl, br = xppm.blbr_iord8(u, al, dm)
-            bl, br = xppm.bl_br_edges(bl, br, u, dxa, al, dm)
 
-            with horizontal(region[i_start + 1, :], region[i_end - 1, :]):
-                bl, br = yppm.pert_ppm_standard_constraint_fcn(u, bl, br)
+@gtscript.function
+def advect_u_along_x(
+    ub_contra: FloatField,
+    u: FloatField,
+    dx: FloatFieldIJ,
+    dxa: FloatFieldIJ,
+    rdx: FloatFieldIJ,
+    dt: float,
+):
+    """
+    Advect covariant x-wind on D-grid using contravariant x-wind on cell corners.
 
-        # Zero corners
-        with horizontal(
-            region[i_start - 1 : i_start + 1, j_start],
-            region[i_start - 1 : i_start + 1, j_end + 1],
-            region[i_end : i_end + 2, j_start],
-            region[i_end : i_end + 2, j_end + 1],
-        ):
-            bl = 0.0
-            br = 0.0
+    Named xtp_u in the original Fortran code.
```

Comment:
This is not true anymore because you've changed the units of what is passed in and what is calculated (specifically when dt gets applied). You should mention this change in the comment.

Reply:
Added to comment.